### ISSUE TITLE
[HOLD for payment 2024-11-21] [$250] Workspace chat - Report header shows "0.00 for Pending... @1.00 / mi" when changing distance & rate offline

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#49278]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.35-4
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4966766
**Email or phone of affected tester (no customers):** applausetester+kh010901@applause.expensifail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
Precondition:
- Workspace has at least two distance rates.

1. Go to staging.new.expensify.com
2. Go to workspace chat.
3. Submit a distance expense.
4. Go to transaction thread.
5. Go offline.
6. Click Distance.
7. Edit the distance by swapping the waypoints.
8. Save it.
9. Click Rate.
10. Select a different rate.

## Expected Result:
Report header will display "Pending" after changing distance and rate offline (production behavior).

## Actual Result:
Report header displays "0.00 for Pending... @1.00 / mi" after changing distance and rate offline.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/c5aff1a5-907a-4142-95cf-eaba49917a8d

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021835712136231108969</li>
        <li>Upwork Job ID: 1835712136231108969</li>
        <li>Last Price Increase: 2024-11-11</li>
<li>Automatic offers: </li>
<ul>
<li>paultsimura | Contributor | 104755241</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by mountiny** [comment]
This is very minor and edge case flow with no real burden to the user, demoting

**Comment by mountiny** [comment]
$50 for finding the offending PR

**Comment by CyberAndrii** [comment]
Offending PR seems to be https://github.com/Expensify/App/issues/47136

**Comment by s77rt** [comment]
cc @paultsimura Can you check this one please

**Comment by alexpensify** [comment]
@s77rt do you agree with this plan? Thanks!

**Comment by s77rt** [comment]
@neil-marcellini is working on https://github.com/Expensify/App/issues/43588 and this one is not a priority, holding makes sense to me

**Comment by alexpensify** [comment]
Thanks for the update. I've moved this to hold, and we can revisit it after #43588 is in production.

**Comment by alexpensify** [comment]
Weekly Update: On hold for https://github.com/Expensify/App/issues/43588

**Comment by alexpensify** [comment]
@s77rt - the update is live in NewExpensify, but one is pending for Old. Should we continue to wait or move forward here? Thanks!

**Comment by s77rt** [comment]
@alexpensify We still need to wait for the frontend changes (https://github.com/Expensify/App/pull/50001)

**Comment by alexpensify** [comment]
Weekly Update: Updated on hold for https://github.com/Expensify/App/pull/50001

**Comment by alexpensify** [comment]
@s77rt the changes in https://github.com/Expensify/App/pull/50001 went to production yesterday. Can we move forward now?

**Comment by s77rt** [comment]
@alexpensify Let's remove the hold. We now have two bugs: 

1. If we simply swap the waypoints the distance does not change to "Pending" but it should because the roadmap can still change

https://github.com/user-attachments/assets/86a458b0-e77b-4283-8c34-f453a15bb2ab

2. The report header still changes and shows "0.00 for Pending" if we add a new waypoint


https://github.com/user-attachments/assets/08d8bf26-6d2d-4c44-b61f-68c9a309b73c

**Comment by s77rt** [comment]
@paultsimura I don't think we need to split [`calculateAmountForUpdatedWaypointOrRate`](https://github.com/Expensify/App/blob/866ce4b3957111d2c3ea5a7dccc3c86b90ed4e7c/src/libs/TransactionUtils/index.ts#L826) into two. The bug is that `hasModifiedRateWithPendingWaypoints` returns `false` because the `updatedTransaction` [overwrites](https://github.com/Expensify/App/blob/866ce4b3957111d2c3ea5a7dccc3c86b90ed4e7c/src/libs/TransactionUtils/index.ts#L326-L338) the existing `pendingFields`.

Simply adding `...updatedTransaction.pendingFields` won't work because after https://github.com/Expensify/App/pull/50001 we added these two lines https://github.com/Expensify/App/blob/be80f3d54bd1f36dd244da680de8140bd31d7de2/src/libs/actions/IOU.ts#L2735 and https://github.com/Expensify/App/blob/be80f3d54bd1f36dd244da680de8140bd31d7de2/src/libs/actions/IOU.ts#L2950

and we don't want to clear pending fields that are set in ACTION1 by ACTION2.

cc @neil-marcellini (for context)

**Comment by s77rt** [comment]
I think we need to revert the above two lines and handle the problem that we were trying to resolve in a different way, the problem was that if you change the rate to a rate that uses a different unit we also need to change the waypoints (the distance) https://github.com/Expensify/App/blob/f0ec762fc6514b4e2627cbad519e6f40a1e6fd97/src/libs/TransactionUtils/index.ts#L286-L292

Now we basically need to move the above logic and add `waypoints` to `transactionChanges`

**Comment by s77rt** [comment]
@paultsimura Thanks for the proposal.

Regarding the first bug, the RCA and the solution look good to me.
Regarding the second bug, I don't think handling both fields in one function is the problem but I see what you are referring to, here is what I think we should do:
1. Move the `calculateAmountForUpdatedWaypointOrRate` logic into `getUpdatedTransaction` so the `updatedTransaction` becomes a constant. In the `getUpdatedTransaction` there is the if-blocks where you can have the logic done depending on what field we are actually changing
2. Move https://github.com/Expensify/App/blob/26e1d2d279cb78d62c31d0c189c7dff5b94e7889/src/libs/actions/IOU.ts#L2544-L2545 and https://github.com/Expensify/App/blob/26e1d2d279cb78d62c31d0c189c7dff5b94e7889/src/libs/actions/IOU.ts#L2666-L2668 to `getUpdatedTransaction`. This will ensure a single source of truth
3. Fix `getUpdatedTransaction` because it's still returning wrong pendingFields (even though this is not necessary to fix our bug here it's still something we should do)

Let me know if that makes sense

**Comment by s77rt** [comment]
@paultsimura We don't necessary have to remove that function, just use it in `getUpdatedTransaction`.

**Comment by s77rt** [comment]
Same ^

**Comment by alexpensify** [comment]
I think we should be able to get proposals now. I updated the title.

**Comment by s77rt** [comment]
@paultsimura Overall looks good to me but can we fix `getUpdatedTransaction` as well? (to return the correct `pendingFields`)

**Comment by s77rt** [comment]
Right, that's another reason that we need to add `...updatedTransaction.pendingFields,`

**Comment by s77rt** [comment]
Still working on the proposals

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
@paultsimura Thank you! Looks good to me but can you check if the linter complains about mutating `transactionChanges`?

**Comment by s77rt** [comment]
@paultsimura Got it! Let's handle this in the PR.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/49278#issuecomment-2435227201)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by CyberAndrii** [comment]
I think I should be paid $50 for [finding the offending PR](https://github.com/Expensify/App/issues/49278#issuecomment-2353782763). Am I right?

**Comment by s77rt** [comment]
# BugZero Checklist:

- [X] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [X] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production
  - [X] 2b. Reported on staging (deploy blocker)
  - [ ] 2c. Reported on both staging and production
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [X] 3d. QA
  - [ ] 3z. Other:

</details>

- [X] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: https://github.com/Expensify/App/pull/47136#discussion_r1845410451 and https://github.com/Expensify/App/pull/25399#discussion_r1845410455

- [X] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: N/A

- [X] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by alexpensify** [comment]
Payouts due: 2024-11-21

- [ ] Contributor: $250 @paultsimura - Paid via Upwork 
- [ ] Reviewer: $250 @s77rt - Please submit a request via Chat 

Upwork job is [here](https://www.upwork.com/jobs/~021835712136231108969).

**Comment by alexpensify** [comment]
No regression test, so no action left here

**Comment by JmillsExpensify** [comment]
$250 approved for @s77rt
